receiver/googlecloudspannerreceiver/metricsreader.go

```diff
@@ -0,0 +1,1282 @@
+
+	if err != nil {
+		return nil, errors.New(fmt.Sprintf("Error occurred during extracting metric values row: %v", err))
+	}
+
+	return metadata.toMetrics(intervalEnd, labelValues, metricValues), nil
+}
+
+func (metadata *MetricsReaderMetadata) toMetrics(intervalEnd time.Time, labelValues []LabelValue,
+	metricValues []MetricValue) []pdata.Metrics {
+
+	var metrics []pdata.Metrics
+
+	for _, metricValue := range metricValues {
+		md := pdata.NewMetrics()
+		rms := md.ResourceMetrics()
+		rm := rms.AppendEmpty()
+
+		ilms := rm.InstrumentationLibraryMetrics()
+		ilm := ilms.AppendEmpty()
+		metric := ilm.Metrics().AppendEmpty()
+		metric.SetName(metadata.MetricNamePrefix + metricValue.getMetricName())
+		metric.SetUnit(metricValue.getMetricUnit())
+		metric.SetDataType(metricValue.getMetricDataType())
+
+		var dataPoints pdata.NumberDataPointSlice
+
+		switch metricValue.getMetricDataType() {
+		case pdata.MetricDataTypeGauge:
+			dataPoints = metric.Gauge().DataPoints()
+		case pdata.MetricDataTypeSum:
+			dataPoints = metric.Sum().DataPoints()
+		}
+
+		dataPoint := dataPoints.AppendEmpty()
+
+		switch valueCasted := metricValue.(type) {
+		case Float64MetricValue:
+			dataPoint.SetDoubleVal(valueCasted.value)
+		case Int64MetricValue:
+			dataPoint.SetIntVal(valueCasted.value)
+		}
+
+		dataPoint.SetTimestamp(pdata.NewTimestampFromTime(intervalEnd))
+
+		for _, labelValue := range labelValues {
+			switch valueCasted := labelValue.(type) {
+			case StringLabelValue:
+				dataPoint.Attributes().InsertString(valueCasted.labelName, valueCasted.value)
+			case BoolLabelValue:
+				dataPoint.Attributes().InsertBool(valueCasted.labelName, valueCasted.value)
+			case Int64LabelValue:
+				dataPoint.Attributes().InsertInt(valueCasted.labelName, valueCasted.value)
+			case StringSliceLabelValue:
+				dataPoint.Attributes().InsertString(valueCasted.labelName, valueCasted.value)
+			case ByteSliceLabelValue:
+				dataPoint.Attributes().InsertString(valueCasted.labelName, valueCasted.value)
+			}
+		}
+
+		dataPoint.Attributes().InsertString(projectIdLabelName, metadata.projectId)
+		dataPoint.Attributes().InsertString(instanceIdLabelName, metadata.instanceId)
+		dataPoint.Attributes().InsertString(databaseLabelName, metadata.databaseName)
+
+		metrics = append(metrics, md)
+	}
+
+	return metrics
+}
+
+/* ------------------------------------------------------------------------------------------------------------------ */
+
+func NewTopQueryStatsMetricsReaderMetadata(
+	projectId string,
+	instanceId string,
+	databaseName string,
+	topMetricsQueryMaxRows int) *MetricsReaderMetadata {
+
+	query := "SELECT * FROM spanner_sys.query_stats_top_minute " +
+		"WHERE interval_end = (SELECT MAX(interval_end) FROM spanner_sys.query_stats_top_minute)" +
+		"ORDER BY EXECUTION_COUNT * AVG_CPU_SECONDS DESC"
+
+	// Labels
+	queryLabelValuesMetadata := []LabelValueMetadata{
+		StringLabelValueMetadata{
+			QueryLabelValueMetadata{
+				labelName:       "query_text",
```

Comment:
Outdated code, it has been already reworked.

Reply:
In case we name field `Name` we also have `Name()` in interface and that is rising compiler error. We want to have good readability until dynamic metric definition will be introduced.